### ISSUE TITLE
Testing the yolo model on a customized training pipeline

### ISSUE THREAD

**Issue by sandeepangara** [ultralytics/ultralytics#22333]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am trying to train the YOLO model in my pipeline by loading it from Ultralytics. Although I changed the num_classes in the detection head, when I perform the forward pass with the data, YOLO still outputs a size similar to the COCO dataset.

def force_refresh_model(model):
    """Clear YOLOv8 DetectionModel caches after head rebuild."""
    if hasattr(model, "stride"):
        delattr(model, "stride")
    if hasattr(model, "names"):
        delattr(model, "names")
    if hasattr(model, "initialize_biases"):
        model.initialize_biases()
    if hasattr(model.model[-1], "stride"):
        model.model[-1].stride = torch.tensor([8., 16., 32.])

    # Re-run dummy forward to rebuild graph
    _ = model(torch.zeros(1, 3, 640, 640))
    print("YOLO model graph reinitialized with new detect head.")
    return model

def get_in_channels(det):
    """Safely extract input channels from each detection head."""
    in_channels = []
    for block in det.cv3:
        if isinstance(block, nn.Sequential) and hasattr(block[0], "in_channels"):
            in_channels.append(block[0].in_channels)
        else:
            for layer in block.modules():
                if isinstance(layer, nn.Conv2d):
                    in_channels.append(layer.in_channels)
                    break
    return in_channels

def rebuild_detect_head(model, num_classes=1):
    det = model.model.model[-1]
    reg_max = getattr(det, "reg_max", 16)
    ch = get_in_channels(det)

    # --- version-safe detect rebuild ---
    try:
        new_det = Detect(nc=num_classes, ch=ch, reg_max=reg_max)
    except TypeError:
        new_det = Detect(nc=num_classes, ch=ch)
        new_det.reg_max = reg_max

    # copy essential attributes
    new_det.stride = det.stride
    new_det.i = det.i
    new_det.f = getattr(det, "f", [-1])   # ensure graph linkage
    new_det.export = det.export
    new_det.bias_init()                   # initialize bias

    model.model.model[-1] = new_det

    print(f"Rebuilt Detect layer: nc={num_classes}, no={new_det.no}, channels={ch}")
    return model

yolo = YOLO("yolov8s.pt")
yolo = rebuild_detect_head(yolo, num_classes=1)
model = force_refresh_model(yolo.model)
#model = yolo.model  # pure nn.Module

x = torch.randn(1, 3, 640, 640)
outs = model(x)
for i, out in enumerate(outs):
    print(f"Head {i}: {out.shape}")

class YOLODataset(torch.utils.data.Dataset):
    def __init__(self, images_dir, labels_dir, img_size=640):
        self.img_files = sorted([os.path.join(images_dir, f) for f in os.listdir(images_dir) if f.endswith((".jpg", ".png"))])
        self.label_files = [os.path.join(labels_dir, Path(f).stem + ".txt") for f in self.img_files]
        self.transform = transforms.Compose([
            transforms.Resize((img_size, img_size)),
            transforms.ToTensor()
        ])

    def __len__(self):
        return len(self.img_files)

    def __getitem__(self, idx):
        img = Image.open(self.img_files[idx]).convert("RGB")
        img = self.transform(img)
        label_path = self.label_files[idx]

        targets = []
        if os.path.exists(label_path):
            with open(label_path) as f:
                for line in f.readlines():
                    cls, cx, cy, w, h = map(float, line.strip().split())
                    targets.append([cls, cx, cy, w, h])
        targets = torch.tensor(targets, dtype=torch.float32)
        return img, targets

def yolo_collate_fn(batch):
    imgs, targets = zip(*batch)
    imgs = torch.stack(imgs, 0)          # batch images [B, 3, H, W]
    return imgs, targets    

# -------------------------------
# 1. Load model (as nn.Module)
# -------------------------------
yolo_wrapper = YOLO("yolov8s.pt")
yolo_wrapper.model.nc = 1
yolo_wrapper.model.names = ["instrument"]
model = yolo_wrapper.model  # torch.nn.Module
model.train()
dataset = YOLODataset(train_img_root, train_lbl_root)

# train_loader = DataLoader(dataset, batch_size=8, shuffle=True, num_workers=4, pin_memory=False)

train_loader = DataLoader(
    dataset,
    batch_size=8,
    shuffle=True,
    num_workers=4,
    pin_memory=False,
    collate_fn=yolo_collate_fn,          # <-- this line fixes the error
)

# -------------------------------
# 3. Loss function & optimizer
# -------------------------------
loss_fn = YOLOv8Loss(num_classes=1)
optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)

for imgs, targets in train_loader:
    print(imgs.shape)  # e.g. torch.Size([8, 3, 640, 640])
    print([t.shape for t in targets])  # e.g. [torch.Size([1, 5]), torch.Size([3, 5]), ...]
    break

# -------------------------------
# 4. Training loop
# -------------------------------
device = "cuda" if torch.cuda.is_available() else "cpu"
model.to(device)

for epoch in range(10):
    model.train()
    total_loss = 0.0

    for imgs, targets in train_loader:
        imgs = imgs.to(device)
        targets = [[t.to(device) for t in target] for target in targets]

        preds = model(imgs)
        for p in preds:
            print(p.shape)  # Each prediction tensor shape

I am getting the following output 

Rebuilt Detect layer: nc=1, no=65, channels=[128, 256, 512]
YOLO model graph reinitialized with new detect head.
Head 0: torch.Size([1, 65, 80, 80])
Head 1: torch.Size([1, 65, 40, 40])
Head 2: torch.Size([1, 65, 20, 20])
The output after forward pass
torch.Size([8, 144, 80, 80])
torch.Size([8, 144, 40, 40])
torch.Size([8, 144, 20, 20])

Any help would be appreciated. Thanks.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sandeepangara, thank you for your interest in Ultralytics 🚀! This is an automated response to help get things moving. An Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including:
-Exact `ultralytics` version and OS/GPU/CUDA/PyTorch versions
-The exact model/checkpoint you’re loading (e.g., `yolov8s.pt`) and any modifications you made
-A short runnable snippet that reproduces the behavior end-to-end
-Console logs showing model summary and forward output shapes
-A few sample images/labels (redacted ok) and your training logs
And verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Why not just rebuild the model using YAML and setting the correct `nc` inside the YAML?

If you want pretrained weights, you can just load the pretrained weights to it.

```
model = YOLO("yolo11n.yaml").load("yolo11n.pt")
```

**Comment by sandeepangara** [comment]
Thanks for the reply. I am looking to pass modify **nc** using a argument and the model is unable to overwrite. 

> Why not just rebuild the model using YAML and setting the correct `nc` inside the YAML?
> 
> If you want pretrained weights, you can just load the pretrained weights to it.
> 
> ```
> model = YOLO("yolo11n.yaml").load("yolo11n.pt")
> ```

**Comment by Y-T-G** [comment]
But what's the reason for not using model YAML to change `nc`?

**Comment by sandeepangara** [comment]
> But what's the reason for not using model YAML to change `nc`?

I want to understand why, after changing the **nc** parameter, the model loads the Coco dataset classes.

**Comment by Y-T-G** [comment]
Changing `nc` by overwriting it doesn't change the head. The head was already created with old `nc`. You're reinitializing head in the first part of your code, which would change the output. But then you just reload the model again from the pretrained checkpoint later, undoing all the changes from the previous code.
